doc/developer/design/20231201_catalog_migration_to_persist.md

```diff
@@ -0,0 +1,164 @@
+- Associated: https://github.com/MaterializeInc/materialize/issues/22392
+
+## The Problem
+
+Currently, we persist all users catalog data in the stash. We would like to move all of that data
+to Persist. This will have the following benefits:
+
+- A step towards architecting a shareable catalog that will contribute to use case isolation.
+- Reduced operational overhead related to maintaining the stash. Note, the overhead is not
+  completely removed because the storage controller still uses the stash.
+
+## Success Criteria
+
+All user catalog data is stored in Persist.
+
+## Out of Scope
+
+- Fully shareable differential catalog.
+- Use case isolation.
+- Zero downtime upgrades.
+
+## Solution Proposal
+
+### Config Collection
+
+The config collection is a subset of catalog data that can never be migrated. As a consequence, this
+data can be read before the catalog is fully opened and can be used to bootstrap an environment. The
+config collection has keys of type `String` and values of type `u64`. Booleans can be modeled using
+this collection by storing a `0` for `false` and `1` for `true`. This proposal will add a key called
+`tombstone`, with a boolean value that indicates if a specific catalog backend is retired. A value
+of `false` means that the backend is not retired (i.e. is still in use) and a value of `true` means
+the backend is retired (i.e. is not still in use).
+
+### Epoch Fencing
+
+The catalog durably stores a fencing token called the epoch. All catalog writers and readers also
+store an in-memory epoch. When a new writer or reader connects to the catalog they store the
+previous epoch plus one in memory and persist this new value durably. Every time a reader or write
+performs a read or write, they compare their in-memory token with the persisted token, and if
+there's a difference, the operation fails. As a result, whenever a reader/write increments the epoch
+durably, they are effectively fencing out all previous readers/writers.
+
+### Catalog Open
+
+Opening a catalog involves the following two steps:
+
+1. Increment the epoch.
+2. Either initialize the persisted catalog state if it's uninitialized, or migrate the data format
+   if there are any migrations, otherwise do nothing.
+
+### Migrate From Stash to Persist
+
+Below are the steps to migrate from the stash to persist.
+
+1. Open the stash catalog.
+2. Open the persist catalog.
+3. If the stash tombstone is `true`, then we're done.
+4. Read stash catalog snapshot.
+5. Replace the contents of the persist catalog with the stash catalog snapshot.
+6. Set the stash tombstone to `true`.
+
+### Rollback From Persist to Stash
+
+Below are the steps to rollback from persist to the stash.
+
+1. Open the stash catalog.
+2. Open the persist catalog.
+3. If the stash tombstone is `false` or doesn't exist, then we're done.
+4. Read persist catalog snapshot.
+5. Replace the contents of the stash catalog with the persist catalog snapshot.
+6. Set the stash tombstone to `false`.
```

Comment:
Yeah, if we crash between steps 5 and 6 we should be fine. When writing this I actually tried to add a specific section on that scenario, but couldn't figure out the right words. 

The idea is that at all times one of the backing stores is the "source of truth", i.e. contains correct data, and the other backing store can be treated as having complete garbage. It might have stale data, it might have no data, it might even have the correct up to date data, but we just treat it as complete garbage. So after step 5, both the stash and persist have the exact same correct data. However, since we haven't flipped the tombstone flag, we treat the stash data as the source of truth and the persist data as garbage. If we crash here, we have the start the whole process over, as if we had never copied the data into persist, even if we end up making no real changes. The same idea applies if another writer fences us out in-between step 5 and 6, they have to start from the beginning and treat the persist data as garbage and the stash as the source of truth.